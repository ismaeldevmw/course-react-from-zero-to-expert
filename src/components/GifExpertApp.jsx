import React, { useState } from 'react'
import { AddCategory } from './AddCategory';

const GifExpertApp = () => {

    // const categories = ['One Punch', 'Samurai', 'Dragón'];
    const [categories, setCategories] = useState(['One Punch', 'Samurai', 'Dragón']);

    // const handleAdd = () => {
    //     // setCategories([...categories, 'HunterXHunter']);
    //     setCategories( cats => [...cats, 'HunterXHunter']);
    // }

    return (
        <>

            <h2>GifExpertApp</h2>
            <AddCategory />
            <hr/>

            <ol>
                {
                    categories.map( (category,  i) => {
                        return <li key={ i }>{ category }</li>
                    })
                }
            </ol>
        </>
    )
}

export default GifExpertApp
