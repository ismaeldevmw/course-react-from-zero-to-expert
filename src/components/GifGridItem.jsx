import React from 'react';
import PropTypes from 'prop-types';

export const GifGridItem = ({ id, title, url }) => {    

    return (
        <div className="card animate__animated animate__fadeIn">
            <div className="card__body">
                <img src={ url } alt={ title} />
                <p>{ title }</p>
            </div>
        </div>
    )
}

GifGridItem.propTypes = {
    title: PropTypes.string.isRequired,
    url: PropTypes.string.isRequired
}
