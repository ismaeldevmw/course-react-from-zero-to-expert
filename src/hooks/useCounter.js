import { useState } from "react"

export const useCounter = ( initialState = 10 ) => {

    const [counter, setCounter] = useState(initialState);

    const increment = ( factor = 1 ) => {
        setCounter( counter + factor );
    }

    const decrease = ( factor = 1) => {
        setCounter( counter - factor);
    }

    const reset = () => {
        setCounter(initialState);
    }


    return {
        counter,
        increment,
        decrease,
        reset
    };
}
