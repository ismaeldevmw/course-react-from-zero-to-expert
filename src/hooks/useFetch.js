import { useEffect, useRef, useState } from 'react';


export const useFetch = ( url ) => {
    
    const isMounted = useRef(true);
    const [state, setState] = useState({ data: null, loading: true, error: null })

    useEffect(() => {
        
        return () => {
            isMounted.current = false;
        }
    }, [])

    useEffect(() => {

        fetch( url )
            .then( resp => resp.json() )
            .then( data => {

                if ( isMounted.current ) {
                    setState({
                        loading: false,
                        error: null,
                        data
                    })
                }
            
            })
            .catch( () => {
                setState({
                    data: null,
                    loading: false,
                    error: 'Can`t loading the information'
                })
            })
            
        return () => {
            setState({ data: null, loading: true, error: null });
        };

    }, [url])

    return state;

}
