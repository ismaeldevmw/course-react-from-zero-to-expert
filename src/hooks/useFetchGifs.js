import { useState, useEffect } from "react";
import { getGifs } from '../helpers/getGifs';

export const useFetchGifs = ( category ) => {
    
    const [state, setState] = useState({
        data: [],
        loading: true
    });

    // It allow us to run once when the page is rendered for the first time
    useEffect( () => {
        
        getGifs( category )
            .then( images => {
                setState({
                    data: images,
                    loading: false
                });
            });
            
    }, [category]);

    return state; // { data: [], loading: true }
}
