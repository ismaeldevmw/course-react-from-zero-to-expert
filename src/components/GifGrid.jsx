import React, { useState, useEffect } from 'react';
import { GifGridItem } from './GifGridItem';
import { getGifs } from '../helpers/getGifs';

export const GifGrid = ({ category }) => {

    const [images, setImages] = useState([]);
    
    // It allow us to run once when the page is rendered for the first time
    useEffect( () => {
        getGifs( category )
            .then( setImages);
    }, [ category ]);
    
    return (
        <>
            <h3> { category } </h3>
            <div className="card-grid">
                {
                    images.map( img => (
                        <GifGridItem 
                            key={ img.id }
                            { ...img }
                        />
                    ))
                }          
            </div>
        </>
    )
}
