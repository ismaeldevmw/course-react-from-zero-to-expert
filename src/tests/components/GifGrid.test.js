import React from 'react';
import { shallow } from 'enzyme';
import { GifGrid } from '../../components/GifGrid';
import { useFetchGifs } from '../../hooks/useFetchGifs';
jest.mock('../../hooks/useFetchGifs');

describe('Tests into GifGrid component', () => {
    
    
    test('should show the component alright', () => {
        
        useFetchGifs.mockReturnValue({
            data: [],
            loading: true
        });
        
        const wrapper = shallow(<GifGrid category="One Punch" />)
        expect(wrapper).toMatchSnapshot();
        
    });
    
    test('should show items when load imagen with useFetchGifs', () => {

        const gifs = [{
            id: 'ABC',
            url: 'https://localhosy/anything/thing.jpg',
            title: 'Anything'
        }]

        useFetchGifs.mockReturnValue({
            data: gifs,
            loading: false
        });
        
        const wrapper = shallow(<GifGrid category="One Punch" />);

        expect( wrapper.find('p').exists() ).toBe(false);
        expect( wrapper.find('GifGridItem').length ).toBe( gifs.length );

    });
});