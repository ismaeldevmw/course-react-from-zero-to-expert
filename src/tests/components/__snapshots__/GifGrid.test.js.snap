// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests into GifGrid component should show the component alright 1`] = `
<Fragment>
  <h3>
     
    One Punch
     
  </h3>
  <p
    className="animate__animated animate__bounce"
  >
    Loading...
  </p>
  <div
    className="card-grid"
  />
</Fragment>
`;
